content/authentication/reading/hashing-passwords/_index.md

```diff
@@ -0,0 +1,148 @@
+
+```csharp {linenos=table}
+   string hash = h("lover1989"); // returns "bd62d7e13ef465fa705f30de198ba0cb"
+
+```
+
+Then we store Jamie's username along with her hashed password in our `user` table.
+
+| username | passwordHash |
+|-----|-----|
+| tswizzle_fan | bd62d7e13ef465fa705f30de198ba0cb |
+
+This is very secure. Even if somebody breaks into the database and finds Jamie's info, they will not be able to log in. Since we used a hash function, the hacker will have a very hard time turning the hash into a password:[see property 2](http://localhost:1313/authentication/reading/hashing-passwords/#hash-properties)
+
+We can still authenticate Jamie, however. When they come to our site to log in, they will submit a username and password. Let's call the password value `submittedPassword`. Some basic logic will allow us to determine, with an *extremely* high rate of probability, whether or not the pair is valid.
+
+To check Jamie's username/password pair, we could do something like this: 
+
+```csharp {linenos=table}
+// fetches Jamie's user object from the database
+User user = GetUserByUsername("tswizzle_fan");
+
+// gets the hash value stored on their object
+string passwordHash = user.PasswordHash;
+
+// hashes the submitted password
+string submittedHash = h(submittedPassword);
+
+if (passwordHash == submittedHash)
+{
+   // the hashes are the same, the passwords can be assumed to be the same
+}
+else
+{
+   // the hashes are different, so the passwords are definitely different
+}
+```
+
+<!-- TODO: Links will need to be changed for the below hash properties -->
+The conditional compares the values of the hash stored in the database with the hash generated from the *submitted* password. By [hash property 1](http://localhost:1313/authentication/reading/hashing-passwords/#hash-properties), we know that if the hash values are different, then there is no way the passwords are the same. By [hash property 3](http://localhost:1313/authentication/reading/hashing-passwords/#hash-properties), we can safely assume that the passwords are the same. 
+
+{{% notice blue "Note" "rocket" %}}
+When using Identity, the library handles hashing passwords for newly registered users and comparing hashes when logging in a user. The example above is just an example and meant to illustrate what is going on under the hood of Identity.
+{{% /notice %}}
+
+### Hashing Isn't Perfect
+
+Using hash functions to process passwords is not a cure-all. One vulnerability is the possibility for **collisions**. A collision occurs when two different messages have the same hash value. By [hash property 3](http://localhost:1313/authentication/reading/hashing-passwords/#hash-properties), this is supposed to be rare. However, if a collision is found for a given hash function, then it may be possible to create an algorithm to *generate* collisions. In other words, given a specific hash value, the algorithm could generate a string with the *same* hash value.
```

Comment:
```suggestion
Using hash functions to process passwords is not a cure-all. One vulnerability is the possibility for **collisions**. A collision occurs when two different messages have the same hash value. By [hash property 3](#hash-properties), this is supposed to be rare. However, if a collision is found for a given hash function, then it may be possible to create an algorithm to *generate* collisions. In other words, given a specific hash value, the algorithm could generate a string with the *same* hash value.
```